Update tut-integration-ui.md (#1617)
demisto/content-docs: docs/tutorials/tut-integration-ui.md

    - [External schema support](../incidents/incident-classification-mapping#classify-using-a-classification-key)
    - [Can sync mirror in](../integrations/mirroring_integration)
    - [Can sync mirror out](../integrations/mirroring_integration)
    - [Long running integration](../integrations/long-running)

    **NOTE**: Mirroring is not supported in Cortex XSIAM. 

### Parameters

These parameters can be used across all commands in the integration. Some common parameters include the API key used for communicating with the product, your username, whether to use proxy, etc.

For the **Yoda Speak** integration, you want to include the API key and proxy settings, and to allow for insecure requests. You will also include a URL parameter, which tells the integration where to send requests. For each parameter, include a display name that tells the user why the value is used.

***API Key***

> Add the following values for the apikey parameter:

Reviewer: ```suggestion
Add the following values for the API Key parameter:
```

| ------------- |-------------|
| Parameter name    |apikey     |
| Type dropdown      |Authentication     |
| Mandatory      | not selected     |
| Display password      |API Key

In this integration, the free service does not require an API key, and allows up to 60 API calls a day with up to 5 calls an hour. If you require more API calls, [FunTranslations](https://funtranslations.com/) offers a paid service that you can access with an API key.  In this case, you provide the option for users with a paid subscription to enter their API key, but you don’t make it mandatory.  The Authentication parameter type enables you to use Cortex XSOAR/XSIAM’s built-in credential management system to save and use the API key, and ensures that the API key is not displayed to the user and not stored in logs. 

**NOTE:** Many integrations that connect to third party services require an API key for authentication, which is sent with every request to the third party service. Since it’s used by every command that performs an API call, you add it as a global parameter and not an argument. 

![apikey](/doc_imgs/tutorials/tut-integration-ui/apikey_param.png) 
 
***URL***

> Add the following values for the url parameter:

Reviewer: ```suggestion
Add the following values for the URL parameter:
```

**NOTE:** The sample code uses standard Python error handling mechanisms, such as *try*. For more information about errors and exceptions in Python, see the [Python documentation](https://docs.python.org/3/tutorial/errors.html).
In this integration code, you want to raise exceptions when errors occur. The convention is to have a main *try*/*except* block on *main()*, that catches errors and calls *return_error*.

The *return_error* function ensures that playbooks calling these functions will fail and stop, alerting the user to a problem. In integrations and automations, you refrain from calling *return_error* in other places in the code.

### Import

At the beginning of the code, you can import Python libraries, so that its commands are available for our integration. Every integration runs inside a Docker image, and our standard Docker image includes most of the common packages, such as JSON and collections. In our **Yoda Speak** integration, you don’t need to import any libraries, as it only uses the *BaseClient* class, implicitly imported from *CommonServerPython*.

> **NOTE:** When working within a traditional IDE, such as PyCharm, or Visual Studio Code, we recommend [importing the following](../integrations/debugging#debugging-using-your-ide) at the top of your code for debugging purposes.

Reviewer: ```suggestion
**NOTE:** When working within a traditional IDE, such as PyCharm or Visual Studio Code, we recommend [importing the following](../integrations/debugging#debugging-using-your-ide) at the top of your code for debugging purposes.
```

```python
import demistomock as demisto
from CommonServerPython import * 
from CommonServerUserPython import * 
```

For Cortex XSOAR, if you want to use Python libraries that are not included in the standard Cortex XSOAR Docker image, you can [create a customized Docker image](../integrations/docker). For Cortex XSOAR 8, see [Change the Docker Image](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/8/Cortex-XSOAR-Cloud-Documentation/Change-the-Docker-image-in-an-integration-or-script).

### Disable Secure Warnings

> Next you want prevent Python from raising a warning when accessing resources insecurely.

Reviewer: ```suggestion
Next you want to prevent Python from raising a warning when accessing resources insecurely.
```

        return self._http_request(method='POST', url_suffix='yoda', data={'text': text}, resp_type='json',  ok_codes=(200,))
```

The Client is an object that communicates with the API. You create a class called *Client*.

> When a Client object is created, it instantiates a parent *BaseClient* using the params you have set up (whether to use proxy, whether to allow insecure connections, and the base URL). If the user provides values to the *api_key* parameter, the Client sets the headers it will use accordingly. 

Reviewer: ```suggestion
When a Client object is created, it instantiates a parent *BaseClient* using the parameters you have set up (whether to use proxy, whether to allow insecure connections, and the base URL). If the user provides values to the *api_key* parameter, the Client sets the headers it will use accordingly. 
```

    - *outputs_prefix*:  The first level of the output in the context data. It usually matches the name of the integration or service. 
    - *raw_response*: The argument used to attach the raw response received from the service, which can be useful when debugging unexpected behaviors. 
    - *outputs_key_field*: Since you can run the translation command multiple times, and possibly receive different results for the same string of text, the system needs to know where to update or append each result. In this example you tell the system that *Phrase.Original* is the key that represents the original text you translated, so that the next time the command is run on the same string of text, the translated values will update. Learn more about storing results as [context data](../concepts/concepts#context-data).
    - *readable_output*: This is what users see in their War Room when calling the command, so it should be formatted. You can use the *tableToMarkdown* function (from CSP) to turn the JSON into a user-friendly table. You provide *tableToMarkdown* with both the JSON values and a title for the table. 
    

>         **NOTE:** The Script Helper provides an easy way to insert common functions into your code. If you click on the **Script Helper** button and search for the *tableToMarkdown* command, you have the option to insert it directly into the code with placeholders for its *name* (title) and *t* (JSON) arguments.

Reviewer: ```suggestion
        **NOTE:** The Script Helper provides an easy way to insert common functions into your code. If you click the **Script Helper** button and search for the *tableToMarkdown* command, you have the option to insert it directly into the code with placeholders for its *name* (title) and *t* (JSON) arguments.
```

```

- ***test-module***
If the command name is ***test-module***, it means the user has clicked the integration test button while setting up or editing an integration instance. **NOTE:** You did not explicitly create a command called ***test-module***. It is a built-in command.
<br /><br />When returning **ok**, the user is shown a green **Success** message. If any value other than **ok** is returned, an error is displayed. Ensure you return errors that help the user understand what to change in the integration settings to fix connection issues. 

- ***yoda-speak-translate***
This is the primary command for our integration and lets us translate strings of text.

> There is also an *else* option. This returns an error if someone tries to run a command that was created in the YAML  file but does not exist in the Python (PY) file. For example, if you added a command ***yoda-interpret*** in the integration settings, but did not add it to this file, and then tried to run that command, you would see **Yoda-interpret is not implemented**.

Reviewer: ```suggestion
There is also an *else* option. This returns an error if someone tries to run a command that was created in the YAML file but does not exist in the Python (PY) file. For example, if you added a command ***yoda-interpret*** in the integration settings, but did not add it to this file, and then tried to run that command, you would see **yoda-interpret is not implemented**.
```

if __name__ in ('__main__', '__builtin__', 'builtins'):
    main()
```

> This line tells the system where to start running our code. By convention, you call the main function *main*.

Reviewer: ```suggestion
This line tells the system where to start running your code. By convention, you call the main function *main*.
```

 
    ![playbook-quiet](/doc_imgs/tutorials/tut-integration-ui/playbook-quiet.png)

7. Save the playbook.

## Create an Incident

To test our new playbook, go to the **Incidents** page and create a new incident. In the **Details** field, enter the text for translation: *The prequel movies are more entertaining than the new Disney movies*. Assign the **Yoda Speak** playbook, and click **Create new incident**.

For Cortex XSOAR:

![new-incident](/doc_imgs/tutorials/tut-integration-ui/new-incident.png)

For Cortex XSIAM:

![xsiam-incident](/doc_imgs/tutorials/tut-integration-ui/xsiam-incident.png)

Select the incident you just created. Go to the **Work Plan** page and you can see that your playbook executed successfully. Go to the **War Room** and view the translation.

![playbook-complete](/doc_imgs/tutorials/tut-integration-ui/war-room-entry.png)

## We’re Done!	

Your example integration is now complete, and you can use it throughout Cortex XSOAR/XSIAM.

Real world integrations are usually more complex than our example. Like any code, integrations require maintenance and can be extended over time with new features, commands etc.

> To ensure integrations perform as expected, packs can have [unit tests](../integrations/unit-testing), and [test playbooks](../integrations/test-playbooks). Learn more about [contributing content](../contributing/contributing). 

Reviewer: ```suggestion
To ensure integrations perform as expected, content packs can have [unit tests](../integrations/unit-testing), and [test playbooks](../integrations/test-playbooks). Learn more about [contributing content](../contributing/contributing). 
```

## Resources

The [YodaSpeak](https://github.com/demisto/content/tree/master/docs/tutorial-integration/YodaSpeak) content pack is available on GitHub. The pack contains the standard pack file structure and can be viewed as a resource when creating new integrations. 

> Go to **Settings > Integrations**. Click the blue button **BYOI** in the top right corner, and the built-in Cortex XSOAR IDE will open. If you don’t see this button, it means you don’t have the correct permissions required for creating new integrations. Please reach out to your admin for assistance. 

Reviewer: ```suggestion
Go to **Settings > Integrations**. Click the blue button **BYOI** in the top right corner, and the built-in Cortex XSOAR IDE will open. If you don’t see this button, it means you don’t have the correct permissions required for creating new integrations. Ask your admin for assistance. 
```
